### PR TITLE
Copy contrib's storage interface to core

### PR DESCRIPTION
**Pull request by pmm-sumo**
**Description:** As discussed during the SIG, we want to move storage extension to core, starting with the interface (this PR) so persistent buffer implementation (https://github.com/open-telemetry/opentelemetry-collector/issues/2285) could use it

**Link to tracking Issue:** #3424 

**Testing:** Just the interface, no tests

**Documentation:** README.md with API

cc @djaglowski @tigrannajaryan

### CODE REVIEW

**Review by djaglowski — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
Before we review/merge this let's make sure we are happy about the direction that https://github.com/open-telemetry/opentelemetry-collector/pull/3274 goes into. I want to make sure the disk buffering use case is nicely handled by this storage interface.

**Review by carlosalberto — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@pmm-sumo is this up-to-date and ready for review?

File: extension/storage/storage.go
Comment (pmm-sumo) [inline extension/storage/storage.go:75]: There's a suggestion from @pmalek-sumo on the other [PR in contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/4145#discussion_r673740836) to make it a specific type, which I think is a good idea

**Comment by pmm-sumo** [comment]
> @pmm-sumo is this up-to-date and ready for review?

@tigrannajaryan there's one suggestion I mentioned above (and going to apply it) and other than that it's up-to-date. API + implementation is also available in [opentelemetry-collector-contrib PR#4145](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/4145) (there are several strategies how we might apply those changes altogether, one is that we could merge API first, then update contrib's core version, extract implementation into a separate PR and merge it)

File: extension/storage/storage.go
Comment (tigrannajaryan) [inline extension/storage/storage.go:30]: Can we add parameter names to make clear what they are? The last `string` is particularly unclear.

File: extension/storage/storage.go
Comment (tigrannajaryan) [inline extension/storage/storage.go:46]: It was a bit confusing to read, perhaps parenthesis will help:
```suggestion
	// specified key. It should return (nil, nil) if not found
```

File: extension/storage/storage.go
Comment (tigrannajaryan) [inline extension/storage/storage.go:43]: It would be nice to add parameter names to functions. Otherwise it is not immediately clear what they are (e.g. what is a `string` attribute?).

**Review by tigrannajaryan — APPROVED** [review]
LGTM, just a couple minor comments.

**Comment by tigrannajaryan** [comment]
@pmm-sumo I will merge after you respond/resolve comments (non of my comments are blocking).

File: extension/storage/storage.go
Comment (pmm-sumo) [reply extension/storage/storage.go]: I described the last param as "signalName", since that was the initial intention, though it can be more generic and describe any unique client for a given component. I didn't want to make it too complex and left it as "signalName" but perhaps we should be more general about it?

Comment (tigrannajaryan) [reply extension/storage/storage.go]: Yes, I think something more generic would be more suitable. Perhaps `storageName` or `substorageName` and then explain in the comments that each component can have multiple storages and that's why we have storage names?

Comment (pmm-sumo) [reply extension/storage/storage.go]: Yeah, that's a good idea. Updated it with `storageName`

**Comment by tigrannajaryan** [comment]
@pmm-sumo thanks for patience!

**Pull request merged**
